### PR TITLE
Visualization

### PR DESCRIPTION
**Pull request by khughitt**
Added some initial logic for generating visualizations for raw and filtered RNA-Seq data.

I attempted to add a hook for the script in `normed_counts.snakefile`, although this may not be the most appropriate place for it.

There are two main files:
1. reports/templates/visualize_rnaseq.Rmd
2. tools/shared/visualization.R

The first one is a knitr template for visualizing and summarizing RNA-Seq samples. Right now it's pretty basic and just includes a few basic statistics and visualizations, but it should be pretty easy to extend in the future as we see fit.

The second file is a light-weight wrapper script which handles the actual knitting of the report.

I tried to the keep the code generic enough so that the same template can be used to visualize the RNA-Seq data at each stage of processing.

The code has not been completely tested in the context of the pipeline yet so some modifications (e.g. how filepath to template, input and output is specified) may be required.

Feel free to modify how the report generation is hooked into Snakemake; for now I just went with the seemingly simplest approach.

Once this is integrated and working, I will extend it to handle the filtered and normalized RNA-Seq data (some modifications will be needed for instance to decide when to log the data before visualizing). Afterwards I can start to tackle visualization of other components, as well as the final output.

### CODE REVIEW

**Comment by khughitt** [comment]
@daler You should get notified already, but just in case... 

Let me know what your thoughts are on how this should be integrated. You can either accept and modify yourself, or just let me know how you want to change things before merging the PR.

**Comment by seandavi** [comment]
@khughitt, not sure what you were thinking about parameter binding for the R markdown from R, but just in case you hadn't seen docs on parameterized reports, here are the details:

http://rmarkdown.rstudio.com/developer_parameterized_reports.html

Another way to do variable substitution from R is with `brew` before calling `rmarkdown::render`.

File: features/normed_counts.snakefile [hunk 1]
```diff
@@ -15,7 +15,13 @@ rule rnaseq_counts_matrix:
         )
         df.to_csv(output[0], sep='\t')
 
-
+rule rnaseq_visualization_raw:
+    input: '{prefix}/raw/rnaseq_expression/counts_matrix.tsv'
+    output: '{prefix}/reports/raw_rnaseq.html'
+    shell:
+        """
+        {Rscript} tools/rnaseq_data_preparation.R {input} {output} 'raw'
```
Comment (daler) [inline features/normed_counts.snakefile:23]: Should call the new script; fixed in a664104

File: reports/templates/visualize_rnaseq.Rmd [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+3. Add support for coloring samples by a specified field.
+-->
+
+```{r, echo=FALSE, message=FALSE}
+library('knitr')
+library('ggplot2')
+library('gplots')
+library('reshape2')
+library('RColorBrewer')
+source('shared/visualization.R')
```
Comment (daler) [inline reports/templates/visualize_rnaseq.Rmd:15]: Path needs to be relative to where this Rmd file lives; fixed in 1f0a738

File: reports/templates/visualize_rnaseq.Rmd [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+#### Sample heatmap (pearson correlation)
+
+```{r sample_heatmap_cor, echo=FALSE}
+heatmap.2(cor(mat), trace='none')
+```
+
+#### Sample Heatmap (spearman correlation)
+
+```{r sample_heatmap_cor, echo=FALSE}
```
Comment (daler) [inline reports/templates/visualize_rnaseq.Rmd:108]: chunk name conflicts with previous; fixed in aba03ba

**Comment by daler** [comment]
722e7bfeb4 is really all it took to integrate the raw rnaseq report into the pipeline.

So this is now working at the "feature prep" part of the pipeline. It would be nice to generate a report for filtered data at the "run" level, since each run can define its filtering strategy independently.

Currently this would be towards the end of the main `Snakefile`, after the `do_filter` rule. For example, a rule like this might do the trick:

``` python
rule filtered_normed_counts:
    input: '{prefix}/runs/{run}/filtered/normed_counts/normed_counts_filtered.tab'
    output: '{prefix}/runs/{run}/filtered/normed_counts/normed_counts_report.html'
    shell:
        """
        {programs.Rscript.prelude}
        {programs.Rscript.path} tools/visualize_rnaseq.R {input} {output} 'raw'
        """
```

Or more generically for any transformation we want of filtered data from `.tab` -> `.html`, tell the R script what kind of data it's getting:

``` python
rule filtered_normed_counts:
    input: '{prefix}/runs/{run}/filtered/{features_label}/{output_label}_filtered.tab'
    output: '{prefix}/runs/{run}/filtered/{features_label}/{output_label}_filtered_report.html'
    shell:
        """
        {programs.Rscript.prelude}
        {programs.Rscript.path} tools/visualize_rnaseq.R {input} {output} '{features_label}'
        """
```

Or even more generically, to turn any `*.tab` into a report sitting in the same directory as the input data:

``` python
rule filtered_normed_counts:
    input: '{base}.tab'
    output: '{base}_report.html'
    shell:
        """
        {programs.Rscript.prelude}
        {programs.Rscript.path} tools/visualize_rnaseq.R {input} {output}
        """
```

 . . . but that's probably overkill generalization since it defers to the report to do the heuristics on how to handle data.

**Comment by khughitt** [comment]
@seandavi Thanks for the suggestions! I've been using knitr and rmarkdown for a while now and hadn't come across the parameterized report functionality before. What I usually tend to do in my own work is to create a separate child "settings.Rmd" with a global list variable containing all of the relevant parameters. That approach works pretty great for most things, but didn't seem like such a good fit here.

 I will check both that and brew out and see if either of those would work well here.

**Comment by khughitt** [comment]
@daler Thanks for the fixes to integrate the visualization code. The way you set it up looks much nicer than what I had. I'll take a stab at the filtered data next.

@seandavi @daler Feel free to let me know if you think of any other useful stats or plots to include. For now I just included a few basic things that I tend to look at when I'm first working with RNA-Seq data, but there is obviously a lot more that could be done.

**Comment by daler** [comment]
(I've found that conversations in PRs can be hard to find later, so I opened up a new issue on this (#20) for better tracking.)

**Pull request merged**